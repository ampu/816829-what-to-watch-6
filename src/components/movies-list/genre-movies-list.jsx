import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import OperationStatus from '../../constants/operation-status';
import {selectGenreMovies} from '../../store/selectors';

import MoviesList from './movies-list';
import SpinnerLoading from '../spinner-loading/spinner-loading';
import Maintenance from '../maintenance/maintenance';

const GenreMoviesList = ({moviesStatus, movies}) => {

  if (moviesStatus === OperationStatus.PENDING) {
    return <SpinnerLoading/>;
  }

  if (moviesStatus === OperationStatus.REJECTED) {
    return <Maintenance/>;
  }

  return <MoviesList movies={movies}/>;
};

GenreMoviesList.propTypes = {
  moviesStatus: PropTypes.oneOf(Object.values(OperationStatus)),
  movies: MoviesList.propTypes.movies,
};

const mapStateToProps = (state) => ({
  moviesStatus: state.moviesStatus,
  movies: selectGenreMovies(state),
});

export default connect(mapStateToProps)(GenreMoviesList);
