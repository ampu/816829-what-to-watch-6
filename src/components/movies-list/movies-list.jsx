import React, {useState, useEffect, useMemo, useCallback} from 'react';
import PropTypes from 'prop-types';

import SmallMovieCard from './small-movie-card';
import ShowMoreButton from '../show-more-button/show-more-button';

const PAGE_LIMIT = 8;

const MoviesList = ({movies}) => {
  const [limit, setLimit] = useState(PAGE_LIMIT);

  useEffect(() => {
    setLimit(PAGE_LIMIT);
  }, [movies]);

  const handleShowMoreButtonClick = useCallback(() => {
    setLimit((currentLimit) => currentLimit + PAGE_LIMIT);
  }, [setLimit]);

  const renderedMovies = useMemo(() => {
    return Array.from({length: Math.min(limit, movies.length)}, (_, index) => {
      const movie = movies[index];
      return <SmallMovieCard key={movie.id} movie={movie}/>;
    });
  }, [movies, limit]);

  return <>
    <div className="catalog__movies-list">
      {renderedMovies}
    </div>
    {limit < movies.length && <ShowMoreButton onClick={handleShowMoreButtonClick}/>}
  </>;
};

MoviesList.propTypes = {
  movies: PropTypes.arrayOf(SmallMovieCard.propTypes.movie),
};

export default MoviesList;
