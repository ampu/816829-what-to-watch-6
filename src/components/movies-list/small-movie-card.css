.small-movie-card__player {
  object-fit: fill;
}
