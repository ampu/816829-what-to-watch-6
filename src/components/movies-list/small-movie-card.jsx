import React from 'react';
import {Link, generatePath} from 'react-router-dom';
import PropTypes from 'prop-types';

import {MainPath} from '../../constants/paths';
import {useTimeoutActivator} from '../../hooks/use-timeout-activator';

import './small-movie-card.css';

const PLAY_TIMEOUT = 1000;

const SmallMovieCard = ({movie = {}}) => {
  const {
    title = ``,
    preview,
    videoPreview,
  } = movie;

  const [isActive, handleMouseEnter, handleMouseLeave] = useTimeoutActivator(PLAY_TIMEOUT);

  return (
    <article className="small-movie-card catalog__movies-card"
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}>

      <div className="small-movie-card__image">
        {!isActive && <img src={preview} alt={title} width="280" height="175"/>}
        {isActive && (
          <video className="small-movie-card__player"
            src={videoPreview} poster={preview}
            width="280" height="175"
            muted autoPlay/>
        )}
      </div>

      <h3 className="small-movie-card__title">
        <Link className="small-movie-card__link" to={generatePath(MainPath.MOVIE, movie)}>{title}</Link>
      </h3>
    </article>
  );
};

SmallMovieCard.propTypes = {
  movie: PropTypes.shape({
    id: PropTypes.string.isRequired,
    title: PropTypes.string,
  }),
  isActive: PropTypes.bool,
};

export default SmallMovieCard;
