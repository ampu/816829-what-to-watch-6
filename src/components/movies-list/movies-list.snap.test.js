import React from 'react';
import {Router} from 'react-router-dom';
import range from 'lodash.range';

import {render} from '@testing-library/react';

import {MoviesList} from './movies-list';
import {createMemoryHistory} from 'history';

const MOVIES_PAGE_LIMIT = 8;
const RANGE_OFFSET = 1;

describe(`MoviesList component`, () => {
  it(`Should render according to the snapshot without any movies in list`, () => {
    const {container} = render(<MoviesList movies={[]}/>);
    expect(container).toMatchSnapshot();
  });

  it(`Should render according to the snapshot with movies count equals to ${MOVIES_PAGE_LIMIT} or less`, () => {
    const movies = range(RANGE_OFFSET, RANGE_OFFSET + MOVIES_PAGE_LIMIT).map((id) => ({
      id: id.toString(),
      preview: `fakePreview`,
      videoPreview: `fakeVideoPreview`,
    }));
    const {container} = render((
      <Router history={createMemoryHistory()}>
        <MoviesList movies={movies}/>
      </Router>
    ));
    expect(container).toMatchSnapshot();
  });

  it(`Should render according to the snapshot with movies count greater than ${MOVIES_PAGE_LIMIT}`, () => {
    const movies = range(RANGE_OFFSET, RANGE_OFFSET + MOVIES_PAGE_LIMIT * 2).map((id) => ({
      id: id.toString(),
      preview: `fakePreview`,
      videoPreview: `fakeVideoPreview`,
    }));
    const {container} = render((
      <Router history={createMemoryHistory()}>
        <MoviesList movies={movies}/>
      </Router>
    ));
    expect(container).toMatchSnapshot();
  });
});
