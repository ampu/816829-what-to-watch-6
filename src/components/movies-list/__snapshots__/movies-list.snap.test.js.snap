// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesList component Should render according to the snapshot with movies count equals to 8 or less 1`] = `
<div>
  <div
    class="catalog__movies-list"
  >
    <article
      class="small-movie-card catalog__movies-card"
    >
      <div
        class="small-movie-card__image"
      >
        <img
          alt=""
          height="175"
          src="fakePreview"
          width="280"
        />
      </div>
      <h3
        class="small-movie-card__title"
      >
        <a
          class="small-movie-card__link"
          href="/films/1"
        />
      </h3>
    </article>
    <article
      class="small-movie-card catalog__movies-card"
    >
      <div
        class="small-movie-card__image"
      >
        <img
          alt=""
          height="175"
          src="fakePreview"
          width="280"
        />
      </div>
      <h3
        class="small-movie-card__title"
      >
        <a
          class="small-movie-card__link"
          href="/films/2"
        />
      </h3>
    </article>
    <article
      class="small-movie-card catalog__movies-card"
    >
      <div
        class="small-movie-card__image"
      >
        <img
          alt=""
          height="175"
          src="fakePreview"
          width="280"
        />
      </div>
      <h3
        class="small-movie-card__title"
      >
        <a
          class="small-movie-card__link"
          href="/films/3"
        />
      </h3>
    </article>
    <article
      class="small-movie-card catalog__movies-card"
    >
      <div
        class="small-movie-card__image"
      >
        <img
          alt=""
          height="175"
          src="fakePreview"
          width="280"
        />
      </div>
      <h3
        class="small-movie-card__title"
      >
        <a
          class="small-movie-card__link"
          href="/films/4"
        />
      </h3>
    </article>
    <article
      class="small-movie-card catalog__movies-card"
    >
      <div
        class="small-movie-card__image"
      >
        <img
          alt=""
          height="175"
          src="fakePreview"
          width="280"
        />
      </div>
      <h3
        class="small-movie-card__title"
      >
        <a
          class="small-movie-card__link"
          href="/films/5"
        />
      </h3>
    </article>
    <article
      class="small-movie-card catalog__movies-card"
    >
      <div
        class="small-movie-card__image"
      >
        <img
          alt=""
          height="175"
          src="fakePreview"
          width="280"
        />
      </div>
      <h3
        class="small-movie-card__title"
      >
        <a
          class="small-movie-card__link"
          href="/films/6"
        />
      </h3>
    </article>
    <article
      class="small-movie-card catalog__movies-card"
    >
      <div
        class="small-movie-card__image"
      >
        <img
          alt=""
          height="175"
          src="fakePreview"
          width="280"
        />
      </div>
      <h3
        class="small-movie-card__title"
      >
        <a
          class="small-movie-card__link"
          href="/films/7"
        />
      </h3>
    </article>
    <article
      class="small-movie-card catalog__movies-card"
    >
      <div
        class="small-movie-card__image"
      >
        <img
          alt=""
          height="175"
          src="fakePreview"
          width="280"
        />
      </div>
      <h3
        class="small-movie-card__title"
      >
        <a
          class="small-movie-card__link"
          href="/films/8"
        />
      </h3>
    </article>
  </div>
</div>
`;

exports[`MoviesList component Should render according to the snapshot with movies count greater than 8 1`] = `
<div>
  <div
    class="catalog__movies-list"
  >
    <article
      class="small-movie-card catalog__movies-card"
    >
      <div
        class="small-movie-card__image"
      >
        <img
          alt=""
          height="175"
          src="fakePreview"
          width="280"
        />
      </div>
      <h3
        class="small-movie-card__title"
      >
        <a
          class="small-movie-card__link"
          href="/films/1"
        />
      </h3>
    </article>
    <article
      class="small-movie-card catalog__movies-card"
    >
      <div
        class="small-movie-card__image"
      >
        <img
          alt=""
          height="175"
          src="fakePreview"
          width="280"
        />
      </div>
      <h3
        class="small-movie-card__title"
      >
        <a
          class="small-movie-card__link"
          href="/films/2"
        />
      </h3>
    </article>
    <article
      class="small-movie-card catalog__movies-card"
    >
      <div
        class="small-movie-card__image"
      >
        <img
          alt=""
          height="175"
          src="fakePreview"
          width="280"
        />
      </div>
      <h3
        class="small-movie-card__title"
      >
        <a
          class="small-movie-card__link"
          href="/films/3"
        />
      </h3>
    </article>
    <article
      class="small-movie-card catalog__movies-card"
    >
      <div
        class="small-movie-card__image"
      >
        <img
          alt=""
          height="175"
          src="fakePreview"
          width="280"
        />
      </div>
      <h3
        class="small-movie-card__title"
      >
        <a
          class="small-movie-card__link"
          href="/films/4"
        />
      </h3>
    </article>
    <article
      class="small-movie-card catalog__movies-card"
    >
      <div
        class="small-movie-card__image"
      >
        <img
          alt=""
          height="175"
          src="fakePreview"
          width="280"
        />
      </div>
      <h3
        class="small-movie-card__title"
      >
        <a
          class="small-movie-card__link"
          href="/films/5"
        />
      </h3>
    </article>
    <article
      class="small-movie-card catalog__movies-card"
    >
      <div
        class="small-movie-card__image"
      >
        <img
          alt=""
          height="175"
          src="fakePreview"
          width="280"
        />
      </div>
      <h3
        class="small-movie-card__title"
      >
        <a
          class="small-movie-card__link"
          href="/films/6"
        />
      </h3>
    </article>
    <article
      class="small-movie-card catalog__movies-card"
    >
      <div
        class="small-movie-card__image"
      >
        <img
          alt=""
          height="175"
          src="fakePreview"
          width="280"
        />
      </div>
      <h3
        class="small-movie-card__title"
      >
        <a
          class="small-movie-card__link"
          href="/films/7"
        />
      </h3>
    </article>
    <article
      class="small-movie-card catalog__movies-card"
    >
      <div
        class="small-movie-card__image"
      >
        <img
          alt=""
          height="175"
          src="fakePreview"
          width="280"
        />
      </div>
      <h3
        class="small-movie-card__title"
      >
        <a
          class="small-movie-card__link"
          href="/films/8"
        />
      </h3>
    </article>
  </div>
  <div
    class="catalog__more"
  >
    <button
      class="catalog__button"
      type="button"
    >
      Show more
    </button>
  </div>
</div>
`;

exports[`MoviesList component Should render according to the snapshot without any movies in list 1`] = `
<div>
  <div
    class="catalog__movies-list"
  />
</div>
`;
